spring-ai-openai/src/test/java/org/springframework/ai/openai/OpenAiMockTestConfiguration.java

```diff
@@ -0,0 +1,311 @@
+ * limitations under the License.
+ */
+
+package org.springframework.ai.openai;
+
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.theokanning.openai.client.OpenAiApi;
+import com.theokanning.openai.service.OpenAiService;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.ai.openai.client.OpenAiClient;
+import org.springframework.ai.openai.client.metadata.support.OpenAiHttpResponseHeadersInterceptor;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.beans.factory.FactoryBean;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.boot.SpringBootConfiguration;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Profile;
+import org.springframework.lang.Nullable;
+import org.springframework.mock.web.MockHttpServletResponse;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.RequestBuilder;
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.mockwebserver.Dispatcher;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+import okio.Buffer;
+import retrofit2.Retrofit;
+import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
+import retrofit2.converter.jackson.JacksonConverterFactory;
+
+/**
+ * {@link SpringBootConfiguration} for {@literal OpenAI's} API using mock objects.
+ * <p>
+ * This test configuration allows Spring AI framework developers to mock OpenAI's API with
+ * Spring {@link MockMvc} and a test provided Spring Web MVC
+ * {@link org.springframework.web.bind.annotation.RestController}.
+ * <p>
+ * This test configuration makes use of the OkHttp3 {@link MockWebServer} and
+ * {@link Dispatcher} to integrate with Spring {@link MockMvc}.
+ *
+ * @author John Blum
+ * @see okhttp3.mockwebserver.Dispatcher
+ * @see okhttp3.mockwebserver.MockWebServer
+ * @see org.springframework.boot.SpringBootConfiguration
+ * @see org.springframework.test.web.servlet.MockMvc
+ * @since 0.7.0
+ */
+@SpringBootConfiguration
+@Profile("spring-ai-openai-mocks")
+@SuppressWarnings("unused")
+public class OpenAiMockTestConfiguration {
```

Comment:
could this be done at a higher level based on AIClient interface vs vendor specific interfaces?  The end goal is easy mocking that should also be portable across model providers.

Reply:
Agreed on making this (potentially) easier for our developers to use in their testing efforts. I will need to think on this more carefully. In the meantime, I was simply appeasing my infrastructure and framework testing needs.

I did call out having a `spring-ai-test` module in our design document as something to help developers with along these lines. I think overtime, especially with a few more implementations of this AI metadata model for different AI providers (e.g. Azure OpenAI and Huggingface) under our belt, we can iron down the reusable testing components.